### DISCUSSION TITLE
Required Body field that accept None value but has no default value

### DISCUSSION THREAD

**Discussion by sector119** [fastapi/fastapi#9603] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
This argument on my endpoint permit None value
Annotated[int | None, Body(embed=True)] = None

But that one - doesn't
Annotated[int | None, Body(embed=True)]
```

**UPD YuriiMotov:**  MRE
```py
from typing import Annotated
from fastapi import FastAPI, Body

app = FastAPI()


@app.post("/api")
def func(a: Annotated[int | None, Body(embed=True)]):
    return {"a": a}


# curl -X 'POST' \
#   'http://localhost:8000/api' \
#   -H 'accept: application/json' \
#   -H 'Content-Type: application/json' \
#   -d '{
#   "a": null
# }'

# {"detail":[{"type":"missing","loc":["body","a"],"msg":"Field required","input":null}]}
```

### Description

I use some middleware that read endpoint arguments and their order matters, I can't put this arg after args without default values because of order.

How can I make it work?

Thanks!

### Operating System

Linux

### FastAPI Version

0.95.2

### Python Version

3.11.3

**Comment by sector119** [comment]
I get example with Query from link below
https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#use-pydantics-required-instead-of-ellipsis

But it doesn't work with Body

```
from typing import Annotated

from fastapi import Body, FastAPI

app = FastAPI()


@app.post("/view")
async def view(item_id: Annotated[int | None, Body()] = ...):
    results = {"item_id": item_id}
    return results
```

**Comment by camiloavil** [comment]
Hi, I don't know if I get rigth the question. But have you tried define a BaseModel class for the body.

https://fastapi.tiangolo.com/tutorial/body-multiple-params/

You can define a default value of None.

class Item(BaseModel):
    name: str
    description: str | None = None
    price: float
    tax: float | None = None


@app.put("/items/{item_id}")
async def update_item(
    item_id: Annotated[int, Path(title="The ID of the item to get", ge=0, le=1000)],
    q: str | None = None,
    item: Item | None = None,
):

**Reply by sector119** [reply]
No, I talk about 3 separate annotated arguments like this, first of them is required, but still accept None value like this page describes https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#required-with-none
Example
```
@app.get("/items/")
async def read_items(q: Annotated[str | None, Query(min_length=3)] = ...):
    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q:
        results.update({"q": q})
    return results
```

Two other arguments should accept None value but could be optional, because they have None default value

```
from typing import Annotated

from fastapi import Body, FastAPI

app = FastAPI()


@app.post("/view")
async def view(
    param1: Annotated[int | None, Body(embed=True)] = ...,
    param2: Annotated[int | None, Body(embed=True)] = None,
    param3: Annotated[int | None, Body(embed=True)] = None,
) -> dict:
    results = {
        "param1": param1,
        "param2": param2,
        "param3": param3
    }

    return results
```

This code accept 

{"param1": 1, "param2": 2, "param3": 3} and {"param1": 1} and {"param1": 1, "param2": null, "param3": null}

but doesn't accept 
{
  "param1": null,
  "param2": 2,
  "param3": 3
}

Here I get
"msg": "field required", "type": "value_error.missing"

**Reply by sector119** [reply]
```
from pydantic import BaseModel, ValidationError


class Model(BaseModel):
    param1: int | None = ...


try:
    m = Model(param1=None)
except ValidationError as e:
    print(e)
else:
    print(m)
```

Works as expected, "param1=None" is printed without any exceptions raised

**Comment by YuriiMotov** [comment]
MRE:
```py
from typing import Annotated
from fastapi import FastAPI, Body

app = FastAPI()


@app.post("/api")
def func(a: Annotated[int | None, Body(embed=True)]):
    return {"a": a}


# curl -X 'POST' \
#   'http://localhost:8000/api' \
#   -H 'accept: application/json' \
#   -H 'Content-Type: application/json' \
#   -d '{
#   "a": null
# }'

# {"detail":[{"type":"missing","loc":["body","a"],"msg":"Field required","input":null}]}
```

**Reply by AlberLC** [reply]
I think my use case is slightly different from the MRE, but it seems to be the same underlying issue. I’m working with a non-embedded JSON body, and I’m trying to make the body itself required while allowing its value to be `null`.

For example:

```python
from typing import Annotated
from fastapi import Body

@app.put('/api')
async def func(a: Annotated[str | None, Body()]):
    return a
```

What I am looking for is that if there is no request body, FastAPI should return `422`. If the JSON body is `null`, it should return `None`. And if the body is a string `"null"`, it should return the string `"null"`.

In my case I am not using `embed=True`, because I don't want the request to be an object such as `{"a": null}`. The JSON body itself should be the value.

The interesting part is that the request really does contain a JSON `null`. If I access the raw request body with Starlette:

```python
await request.body()
# b'null'
```

FastAPI nevertheless returns:

```json
{
  "detail": [
    {
      "type": "missing",
      "loc": ["body"],
      "msg": "Field required",
      "input": null
    }
  ]
}
```

So this doesn't seem to be a problem with the client not sending a body. The body is present and contains the JSON value `null`.

I also don't want to use `= None`, because that changes the semantics: it makes the body optional, so an entirely missing body would also be accepted. For my API, an omitted body and a body containing JSON `null` are two different cases.

So what I need is essentially a required nullable scalar JSON body:

```text
missing body → 422
body = null  → None
body = value → value
```

Is there currently a supported way to express this with `Body()` without using `embed=True` and without making the body optional?

**Comment by vynazevedo** [comment]
@AlberLC For the non-embedded scalar case in your follow-up, you can distinguish those inputs by checking the raw body as well as the parsed value. This is a workaround, rather than a way to express both requirements with `Body()` alone:

```python
from typing import Annotated
from fastapi import Body, FastAPI, Request
from fastapi.exceptions import RequestValidationError

app = FastAPI()

@app.put("/api", openapi_extra={"requestBody": {"required": True}})
async def endpoint(
    request: Request,
    a: Annotated[str | None, Body()] = None,
):
    if not await request.body():
        raise RequestValidationError([
            {
                "type": "missing",
                "loc": ("body",),
                "msg": "Field required",
                "input": None,
            }
        ])
    return a
```

The default lets JSON `null` get through parsing; the raw-body check separately rejects an absent body. So the endpoint still requires a body, even though the parameter has a default. `openapi_extra` makes that requirement explicit in the generated schema too.

Tested with FastAPI 0.141.1 / Pydantic 2.13.5:

| Request body | Result |
| --- | --- |
| no bytes | 422 |
| `null` | 200, JSON `null` |
| `"null"` | 200, JSON string `"null"` |
| `"hello"` | 200, JSON string `"hello"` |
| `123`, `{}`, invalid JSON or whitespace only | 422 |

Also checked that OpenAPI retains the nullable string schema with `requestBody.required: true`. FastAPI still does the JSON decoding and type validation here; the only added validation is presence of body bytes. The schema override uses the documented [OpenAPI extra](https://fastapi.tiangolo.com/advanced/path-operation-advanced-configuration/#openapi-extra) mechanism.

**Comment by nanisadw3** [comment]
Hi @sector119 👋

To make a Body parameter strictly **required** (i.e. the key must exist in the payload) while explicitly allowing its value to be `null` (`None`), you should use Pydantic's `Field(...)` (Ellipsis) with `Body`:

```python
from typing import Annotated, Optional
from fastapi import FastAPI, Body

app = FastAPI()

@app.post("/api")
def submit_data(
    # '...' marks the field as required, while Optional/None allows null
    a: Annotated[Optional[int], Body(..., embed=True)]
):
    return {"a": a}
```

### Behavior:
1. **Passing `{"a": null}`**: ✅ Accepted (returns `{"a": null}`).
2. **Passing `{"a": 42}`**: ✅ Accepted (returns `{"a": 42}`).
3. **Omitting the key `{}`**: ❌ Fails with `422 Unprocessable Entity` (`missing field "a"`).

In Pydantic V2 / FastAPI 0.100+, this can also be written cleanly as:
```python
a: Annotated[int | None, Body(embed=True)] = ...
```
Both tell FastAPI and Pydantic that the JSON key is mandatory, but `None` is a permitted value.
